File: 4.8-bc/go/stage-3-hexagonal/commerce/app/service.go
package app

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/commerce/domain"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/shared/events"
)

type Service struct {
	repo domain.Repository
	bus  *events.Bus
}

func New(repo domain.Repository, bus *events.Bus) *Service {
	return &Service{repo: repo, bus: bus}
}

func (s *Service) Place(ctx context.Context, userID uuid.UUID, totalCents int64) (*domain.Order, error) {
	if totalCents < 0 {
		return nil, domain.ErrInvalidOrder
	}
	o := domain.Order{
		ID:         uuid.New(),
		UserID:     userID,
		TotalCents: totalCents,
		Status:     "pending",
		CreatedAt:  time.Now().UTC(),
	}
	if err := s.repo.Create(ctx, o); err != nil {
		return nil, err
	}
	s.bus.Publish(ctx, domain.OrderPlaced{OrderID: o.ID, UserID: o.UserID, TotalCents: o.TotalCents, At: o.CreatedAt})
	return &o, nil
}

File: 4.8-bc/py/stage-3-hexagonal/app/commerce/app/service.py
"""Commerce use cases — Place order."""
from __future__ import annotations

from datetime import UTC, datetime
from uuid import UUID, uuid4

from app.commerce.domain.order import InvalidOrderError, Order, OrderPlaced
from app.commerce.domain.repository import Repository
from app.shared.events import EventBus


class Service:
    def __init__(self, repo: Repository, bus: EventBus) -> None:
        self._repo = repo
        self._bus = bus

    async def place(self, user_id: UUID, total_cents: int) -> Order:
        if total_cents < 0:
            raise InvalidOrderError()
        o = Order(
            id=uuid4(),
            user_id=user_id,
            total_cents=total_cents,
            status="pending",
            created_at=datetime.now(UTC),
        )
        await self._repo.create(o)
        await self._bus.publish(
            OrderPlaced(
                order_id=o.id,
                user_id=o.user_id,
                total_cents=o.total_cents,
                at=o.created_at,
            )
        )
        return o
